### DISCUSSION TITLE
Literal syntax preferences

### DISCUSSION THREAD

**Discussion by rossabaker** [http4s/http4s#4078] [Ideas]
We have multiple ways to create literal values, which has doubled our work in the Dotty port.

# Existing

## Old

```scala
@ Uri.uri("http://example.com/"); QValue.q(0.8); MediaType.mediaType("application/json") 
res8_0: Uri = Uri(
  Some(Scheme(http)),
  Some(Authority(None, RegName(example.com), None)),
  "/",
  ,
  None
)
res8_1: QValue = QValue(0.8)
res8_2: MediaType = MediaType(application/json)
```

Note that `QValue.q` takes a double, not a `String`.

To use them unqualified is an import-per-type.  To use them qualified is a redundant name.  We might have called them all `.literal` on the companion, but didn't.

We tried to deprecate these in 0.20, but it doesn't emit warnings, and they're in widespread use.  Particularly `uri`.

## New

Requires an import:

```scala
@ import org.http4s.syntax.literals._ 
import org.http4s.syntax.literals._
```

Alternatively:

```scala
@ import org.http4s.implicits._ 
```

These support interpolation of variables in theory, which was a motivation to move to them:

```scala
@ uri"http://www.example.com/"; qValue"0.8"; mediaType"application/json" 
res10_0: Uri = Uri(
  Some(Scheme(http)),
  Some(Authority(None, RegName(www.example.com), None)),
  "/",
  ,
  None
)
res10_1: QValue = QValue(0.8)
res10_2: MediaType = MediaType(application/json)
```

But not yet in practice:

```scala
@ val host = "example.com"; uri"http://${host}/"; 
cmd11.sc:1: exception during macro expansion: 
scala.MatchError: org.http4s.syntax.`package`.literals.http4sLiteralsSyntax(scala.StringContext.apply("http://", "/")) (of class scala.reflect.internal.Trees$ApplyImplicitView)
	at org.http4s.LiteralSyntaxMacros$.singlePartInterpolator(LiteralSyntaxMacros.scala:74)
	at org.http4s.LiteralSyntaxMacros$.uriInterpolator(LiteralSyntaxMacros.scala:30)

val host = "example.com"; val res11_1 = uri"http://${host}/";
                                        ^
Compilation Failed
```

# Discussion

Should we deprecate one style?  When?

**Comment by Daenyth** [comment]
I think the string context style for *literals* is better.

Can you clarify how the ones like `Uri.uri` are also for literals as opposed to just runtime strings?

**Reply by isomarcte** [reply]
@Daenyth 

> Can you clarify how the ones like Uri.uri are also for literals as opposed to just runtime strings?

If I understand your question, it is because `Uri.uri` has a Scala 2.x macro implementation, it only works with literals as of 0.21, not runtime `String` values.

```scala
scala> Uri.uri("http://localhost")
val res0: org.http4s.Uri = http://localhost

scala> val localHostString: String = "http://localhost"
val localHostString: String = http://localhost

scala> Uri.uri(localHostString)
              ^
       error: This method uses a macro to verify that a String literal is a valid URI. Use Uri.fromString if you have a dynamic String that you want to parse as a Uri.
```

**Reply by rossabaker** [reply]
Correct. Both the string context (e.g., `uri"http://www.example.com/"`) and the method (`Uri.uri("http://www.example.com/"`) are macros that accept literals.  The former is capable of some interpolation, but we haven't implemented it.  Interpolation _and_ validation may not be highly compatible goals.

**Reply by rossabaker** [reply]
Note that the import tax will go up modestly if we spin a URI off into cats-uri.  The more projects our literals come from, the more string context imports we need, unless we compose one from disparate projects, which is dubious.
